zcash_client_sqlite/src/wallet/commitment_tree.rs

```diff
@@ -746,15 +757,98 @@ pub(crate) fn truncate_checkpoints(
     Ok(())
 }
 
+pub(crate) fn put_shard_roots<
+    H: Hashable + HashSer + Clone + Eq,
+    const DEPTH: u8,
+    const SHARD_HEIGHT: u8,
+>(
+    conn: &rusqlite::Transaction<'_>,
+    table_prefix: &'static str,
+    start_index: u64,
+    roots: &[CommitmentTreeRoot<H>],
+) -> Result<(), ShardTreeError<Error>> {
+    if roots.is_empty() {
+        // nothing to do
+        return Ok(());
+    }
+
+    // We treat the cap as a DEPTH-SHARD_HEIGHT tree so that we can make a batch insertion of
+    // root data using `Position::from(start_index)` as the starting position and treating the
+    // roots as level-0 leaves.
+    let cap = LocatedTree::from_parts(
+        Address::from_parts((DEPTH - SHARD_HEIGHT).into(), 0),
+        get_cap(conn, table_prefix).map_err(ShardTreeError::Storage)?,
+    );
+
+    let cap_result = cap
+        .batch_insert(
+            Position::from(start_index),
+            roots.iter().map(|r| {
+                (
+                    r.root_hash().clone(),
+                    Retention::Checkpoint {
+                        id: (),
+                        is_marked: false,
+                    },
+                )
+            }),
+        )
+        .map_err(ShardTreeError::Insert)?
+        .expect("slice of inserted roots was verified to be nonempty");
+
+    put_cap(conn, table_prefix, cap_result.subtree.take_root()).map_err(ShardTreeError::Storage)?;
+
+    for (root, i) in roots.iter().zip(0u64..) {
+        // To avoid deserializing the subtree just to annotate its root, we simply cache the downloaded
+        // root alongside of the persisted subtree and reannotate the root, handling conflicts, at the
+        // time that we deserialize the tree.
+        let mut stmt = conn
+            .prepare_cached(&format!(
+            "INSERT INTO {}_tree_shards (shard_index, subtree_end_height, root_hash, shard_data)
+            VALUES (:shard_index, :subtree_end_height, :root_hash, :shard_data)
+            ON CONFLICT (shard_index) DO UPDATE
+            SET subtree_end_height = :subtree_end_height, root_hash = :root_hash",
```

Comment:
When there is a conflict, why isn't `shard_data` updated? The comment after this says that `shard_data` is only used on the insert side of the upsert, but none of the comments appear to explain why.

It's also possible that I am getting confused because "root" is being used in some places to refer to an entire subtree and in others to refer to just the root node (and its hash) of that subtree.

Reply:
If no entry exists at this index, we need to write a valid serialized tree; however, if a tree already exists we don't want to spend the cycles to deserialize and reserialize it just to change the annotation on the root; instead, we just cache the root and then can reannotate the root with that hash when we actually need to use or update the deserialized subtree at some later point.

Reply:
I think I've figured out the source of my confusion. `put_shard_roots` only changes the value of the root node of each subtree in the range, it does not touch the rest of the subtree's data. But the `subtree` crate uses `root` to refer to the entire tree rooted at a given node, so my brain was expecting the subtree's data to be altered by this method.

Reply:
Oh wait, is it better to think of it as we _should_ be updating `subtree_data` as well, but the only part of it that would get updated is the annotation because none of the rest of the tree structure stores the root's information? That would clarify the comment above `stmt`.

Reply:
That's correct; if it didn't require an expensive round-trip through the serialization logic, we'd reannotate the root node with its hash immediately here.